### PR TITLE
swarm, state, swap/swap: query peer balances

### PR DESCRIPTION
**Pull request by mortelli**
This PR improves the _query balances_ functionality so that:

- Balances with specific nodes can be queried with the `swap_balance` RPC call, provided the `enode.ID` is included in the call as its only parameter.
- Balance queries now check disk data as well. This means that data for nodes that haven't been seen in a while will still be displayed.
  - For this, the `Store` interface was extended with a `Keys` method, which retrieves all keys from the underlying `leveldb` structure. You can then fetch each individual value from disk and merge it with the results in memory.
- Both queries (_single balance_ and _all balances_) have been moved from the `bzz` namespace to the `swap` API namespace. It should probably be `accounting` instead, but to know this for sure we should resolve issue #1591 first.

Related issues: #1447, #1591.

### CODE REVIEW

File: swarm.go
Comment (mortelli) [inline swarm.go:509]: I'm not sure why we deleted this on the first place.

It seems to be present on the `master` branch, but not on the `incentives` branch.

File: state/dbstore.go
Comment (holisticode) [inline state/dbstore.go:111]: you could just after this line say
`defer iter.Release` and then at the end just

`return keys, iter.Error()`

File: swap/swap.go
Comment (holisticode) [inline swap/swap.go]: One variation could be to initialize `balances` right at the top to be `s.balances`:

`balances := s.balances`

Then, when we read from file, we would have to check if the keys is already there and not overwrite.
What do you think is best?

Also, we may have to `lock` while reading the balances from memory

File: swap/swap_test.go
Comment (holisticode) [inline swap/swap_test.go]: Let's not ignore the error coming back, even if it is trivial here?

**Review by holisticode — CHANGES_REQUESTED** [review]
I like the idea with the `Keys()` function.

There are some minor things you may want to consider.

File: state/dbstore.go
Comment (mortelli) [reply state/dbstore.go:111]: Done, nice catch!

I've also updated the code so that either `keys` or `err` _have_ to be `nil`.

File: swarm.go
Comment (nonsense) [reply swarm.go:509]: I'm not sure what will happen when we have the same `rpc.API` namespace entry twice - `bzz` with public API, and `bzz` with admin API, but since we have it on `master` this probably works fine. I just hope that the second entry (the one below) doesn't override the methods from `&Info{s.config}` - probably worth verifying.

**Comment by nonsense** [comment]
I'm not comfortable reviewing PRs without tests being run on them, since this is using the `incentives` branch. In general LGTM, but if there are failed tests, they should be addressed before merging to `master`.

**Review by nonsense — APPROVED** [review]

File: swap/swap_test.go
Comment (janos) [inline swap/swap_test.go]: It is better to check for state.ErrNotFound value than the string returned by the Error function.

File: swap/swap.go
Comment (janos) [inline swap/swap.go]: Just as a suggestion, to remove Get prefixes on methods. This is somehow a general practice, but not required if it makes sense to have it. In this cases, I think we can remove it.

File: swap/swap.go
Comment (janos) [inline swap/swap.go]: I assume that there will be a not so large number of peer IDs returned from the state store. Over time, if we do not cleanup "staled" peers, this list is only growing. What would be the accepted number of keys in memory? Would the iterator pattern or even sending over a channel make sense to avoid potential higher memory consumption? Or it is expected that even a high number of returned keys takes not significant amount of memory or makes no pressure to the gc if called frequently (every string here escapes to the heap, I suppose in this case).

File: swap/swap.go
Comment (mortelli) [reply swap/swap.go]: This comment made me see that this function had redundant logic, since the inner `GetPeerBalance` call already checked both memory and disk, and applied the locks as well.

I've now refactored this function into one with 2 steps: get a list of SWAP peers, and then get the balance for each of them. Both of these steps deal with the locks and memory/disk access, and could be called independently.

While I understand it might not be optimal in terms of performance (while getting the list of peers we could be getting their balances as well) I do think it's the cleanest approach, so as to segregate responsibilities.

File: swap/swap.go
Comment (holisticode) [reply swap/swap.go]: The number of entries will depend on how many peers we have a balance with. 

This means if we never limit the amount of peers we connect through the Kademlia, this list can become big. There was a `MaxPeer` constant at some point, do we still have it?

In normal operation the list should be manageable - up to 100 maybe?

However, if there is a lot of fluctuations in the network, specifically if there are attacks (peers who only suck out services up to payment threshold and then disconnect), this list can become pretty huge, and polluted with stale entries. In this case some archiving function or something, based on a timestamp, might be needed (check https://github.com/ethersphere/swarm/issues/1596).

In any case, the iteration would only be used when a user explicitly calls `GetAllBalances`, which is a manual operation and probably not too frequent.

File: swap/swap.go
Comment (holisticode) [reply swap/swap.go]: Is this somewhere in the `go` guidelines? I agree in this case we can remove it (it was there before, added by me, sorry @mortelli )

File: swap/swap_test.go
Comment (mortelli) [reply swap/swap_test.go]: You're absolutely right, fixed.

I also changed the code so that we are no longer ignoring errors in the `swarm/Balance` and `swarm/Balances` functions; they are now returned.

File: swap/swap.go
Comment (mortelli) [reply swap/swap.go]: Thank you @janos for this very relevant observation.

I agree with @holisticode: let's deal with stale data in #1596.

File: swap/swap_test.go
Comment (mortelli) [reply swap/swap_test.go]: Fixed, thanks.

File: swap/swap.go
Comment (mortelli) [reply swap/swap.go]: Done! Thanks.

Comment (janos) [reply swap/swap.go]: > Is this somewhere in the `go` guidelines? I agree in this case we can remove it (it was there before, added by me, sorry @mortelli )

Thanks, I've read it here https://golang.org/doc/effective_go.html#Getters. Seams nice, but not crucial.

File: swap/swap.go
Comment (janos) [reply swap/swap.go]: Thanks for the explanation @holisticode. As it is a manual operation, and not too frequent, it is not critical. And, I agree, @mortelli that it can be handled in some future iteration/pr. I just wanted to raise attention to it. Thanks.

**Review by janos — APPROVED** [review]
LGTM.

File: swarm.go
Comment (mortelli) [reply swarm.go:509]: I tried methods from both `bzz` APIs and it seems to work OK (`info` vs. `has`, `hive`, etc.).

I don't know what would happen if the two APIs had functions with the same names. So I think it's probably best to change the name of one of them at some point.

**Review by holisticode — APPROVED** [review]

**Pull request merged**
